This fixes the discount_ calculation logic and promotion tax inclusiveness calculation

scherddel opened the pull request:
This PR addresses a bug where the timing to consider the is_tax_inclusive parameter was not right when calculating promotion adjustments. The issue was caused by applying the tax_inclusive calculation too late in the process

Changes:

- Refactored is_tax_inclusive calculation logic in the "promotion to adjustment" process to consitently make adjustments "subtotal" of a promotion amount considering the "is_tax_inclusive" flag as all calculations performed to apply promotions are based on item subtotals 
- Refactored adjustments/index.ts for improved readability
- Removed is_tax_inclusive from cart item line adjustements to reduce redundancy and only keep the flag in the promotion object
- Refactored test cases in promotion and cart to correct misscalculations causes by the mentioned issues

**Known Issue:**
This fix reveals an additional issue that requires follow-up work: In tax-inclusive contexts, promotion adjustments are incorrectly calculated using the line item's subtotal rather than the full price.
For example, with an item priced at 119 (tax-inclusive) and 19% tax, the adjustment calculation (of a 119 adjustment) uses 100 (subtotal) as the basis instead of the full 119 price. This results in incorrect promotion amounts and needs to be addressed in a future PR.

Edit: Also fixes the "Known Issue": 
- Adjustments are now subtotals only which is in line with the implementation of the promotion logic (also only applies subtotals)
- Cleaned up some code to increase readability

Edit2: Due to some concerns raised by oliver I have refactored the solution to use the same logic as promotions and the amount of adjustments now the same but considers different totals based on the is_tax_inclusive flag

scherddel committed:
This fixes the discount_ calculation logic

scherddel commented:
Update on the investigation of the "Known issue":

Unit Price (tax-inclusive): 119
Tax: 19%

Applied Promotion (tax-inclusive): 119
```ts
{
    "id": "cart_01K06M5BRWCDZD1QKR5P8VC5TR",
    "currency_code": "eur",
    "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
    "created_at": "2025-07-15T08:45:30.780Z",
    "updated_at": "2025-07-15T08:46:00.881Z",
    "completed_at": null,
    "total": 19,
    "subtotal": 100,
    "tax_total": 3.033613445378151,
    "discount_total": 100,
    "discount_subtotal": 84.03361344537815,
    "discount_tax_total": 15.966386554621849,
    "original_total": 119,
    "original_tax_total": 19,
    "item_total": 19,
    "item_subtotal": 100,
    "item_tax_total": 3.033613445378151,
    "original_item_total": 119,
    "original_item_subtotal": 100,
    "original_item_tax_total": 19,
    "shipping_total": 0,
    "shipping_subtotal": 0,
    "shipping_tax_total": 0,
    "original_shipping_tax_total": 0,
    "original_shipping_subtotal": 0,
    "original_shipping_total": 0,
    "credit_line_subtotal": 0,
    "credit_line_tax_total": 0,
    "credit_line_total": 0,
    "metadata": null,
    "sales_channel_id": "sc_01JP5CGAA7THA9DR0JJ8QP3A34",
    "shipping_address_id": "caaddr_01K06M695HWH4TKW21R5KBB092",
    "billing_address_id": "caaddr_01K06M695HMNB75F97CF1E4QVS",
    "customer_id": "cus_01JQ4944V31GBEPF64GYETNEWR",
    "items": [
        {
            "id": "cali_01K06M5Y17D7VKXZNN89X27YHM",
            "quantity": 1,
            "unit_price": 119,
            "compare_at_unit_price": null,
            "is_tax_inclusive": true,
            "tax_lines": [
                {
                    "id": "calitxl_01K06M6EDRGQFCKEM025233C0J",
                    "description": "MwSt",
                    "code": "",
                    "rate": 19,
                    "provider_id": "system"
                }
            ],
            "adjustments": [
                {
                    "id": "caliadj_01K06M6EF60C1986DGKR6B02GF",
                    "code": "FLY119",
                    "promotion_id": "promo_01JZJTKMBZ4PE46QBSJ59MQ8TV",
                    "amount": 100,
                    "is_tax_inclusive": true
                }
            ],
        }
    ],
    "promotions": [
        {
            "id": "promo_01JZJTKMBZ4PE46QBSJ59MQ8TV",
            "code": "FLY119",
            "is_automatic": false,
            "application_method": {
                "value": 119,
                "type": "fixed",
                "currency_code": "eur"
            }
        }
    ]
}
```

The promotions says 119 but the adjustment only says 100. The reason is that the adjustment is considering the available subtotal of the product which is incorrect in case of tax-inclusiveness. This is the code from:
https://github.com/medusajs/medusa/blob/491b08e0448e3e7d69c09b9516c39f50e2f691a0/packages/core/utils/src/totals/promotion/index.ts#L136

In this case it would probably make sense to set the applicableTotal = lineItemTotal - promotion.applied_value where the lineItemTotal should really be the total and not the subtotal as defined here:
https://github.com/medusajs/medusa/blob/491b08e0448e3e7d69c09b9516c39f50e2f691a0/packages/core/utils/src/totals/promotion/index.ts#L131

scherddel commented:
Ok so more observations:

The problem is that adjustments are applied to subtotals which are without tax. The adjustment amount will be calculated on the promotions application_method here:
https://github.com/medusajs/medusa/blob/491b08e0448e3e7d69c09b9516c39f50e2f691a0/packages/core/utils/src/totals/promotion/index.ts#L142

The problem is that when the subtotal is larger than the promotion value it will take the total of the promotion as a basis, not the subtotal (119 instead of 100). If the subtotal is smaller than the promotion value it will take the smaller subtotal.

I will try the following: 
- Always make sure that promotion adjustments are subtotal (without tax)
- Always apply the subtotal adjustment to the subtotal of the item

scherddel commented:
I fixed how we apply the promotion and took the tax inclusiveness of the promotion into consideration. It looks good now:
Unit price: 1190 (tax inclusive)
Tax: 19%
Promotion: 119 (tax inclusive)
```ts
{
{
    "id": "cart_01K06YNTJ133TCM6H4FKR9ANAJ",
    "currency_code": "eur",
    "email": "asd@asd.de",
    "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
    "created_at": "2025-07-15T11:49:15.970Z",
    "updated_at": "2025-07-15T11:49:34.258Z",
    "completed_at": null,
    "total": 1071,
    "subtotal": 1000,
    "tax_total": 171,
    "discount_total": 119,
    "discount_subtotal": 100,
    "discount_tax_total": 19,
    "original_total": 1190,
    "original_tax_total": 190,
    "item_total": 1071,
    "item_subtotal": 1000,
    "item_tax_total": 171,
    "original_item_total": 1190,
    "original_item_subtotal": 1000,
    "original_item_tax_total": 190,
    "shipping_total": 0,
    "shipping_subtotal": 0,
    "shipping_tax_total": 0,
    "original_shipping_tax_total": 0,
    "original_shipping_subtotal": 0,
    "original_shipping_total": 0,
    "credit_line_subtotal": 0,
    "credit_line_tax_total": 0,
    "credit_line_total": 0,
    "metadata": null,
    "sales_channel_id": "sc_01JP5CGAA7THA9DR0JJ8QP3A34",
    "shipping_address_id": "caaddr_01K06YPCDHA8E3042YAAK70254",
    "billing_address_id": "caaddr_01K06YPCDH18B20Z6GGJGJY6VF",
    "customer_id": "cus_01K06YBN73HPPT0PK5X87SQ1EC",
    "items": [
        {
            "id": "cali_01K06YNZ6JC7BBC8QNJ0BX2QF3",
            "quantity": 1,
            "unit_price": 1190,
            "compare_at_unit_price": null,
            "is_tax_inclusive": true,
            "tax_lines": [
                {
                    "id": "calitxl_01K06YPHP142R2Z4S0148Y3T2T",
                    "description": "MwSt",
                    "code": "",
                    "rate": 19,
                    "provider_id": "system"
                }
            ],
            "adjustments": [
                {
                    "id": "caliadj_01K06YPHQH0BD3H7QEGX0JT9PM",
                    "code": "FLY119",
                    "promotion_id": "promo_01JZJTKMBZ4PE46QBSJ59MQ8TV",
                    "amount": 100,
                    "is_tax_inclusive": true
                }
            ],
            "product": {
                "id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
                "collection_id": null,
                "type_id": null,
                "categories": [
                    {
                        "id": "pcat_01JP5NSGYQG76M8E68WV13Y4EF"
                    }
                ],
  
            }
        }
    ],
    "promotions": [
        {
            "id": "promo_01JZJTKMBZ4PE46QBSJ59MQ8TV",
            "code": "FLY119",
            "is_automatic": false,
            "application_method": {
                "value": 119,
                "type": "fixed",
                "currency_code": "eur"
            }
        }
    ]
}
```

Unit price: 119 (tax inclusive)
Tax: 19%
Promotion: 119 (tax inclusive)
```ts
{
    "id": "cart_01K06YQAPTFWZWTTNBAF9BZZPJ",
    "currency_code": "eur",
    "email": "asd@ad.de",
    "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
    "created_at": "2025-07-15T11:50:05.274Z",
    "updated_at": "2025-07-15T11:50:22.146Z",
    "completed_at": null,
    "total": 0,
    "subtotal": 100,
    "tax_total": 0,
    "discount_total": 119,
    "discount_subtotal": 100,
    "discount_tax_total": 19,
    "original_total": 119,
    "original_tax_total": 19,
    "item_total": 0,
    "item_subtotal": 100,
    "item_tax_total": 0,
    "original_item_total": 119,
    "original_item_subtotal": 100,
    "original_item_tax_total": 19,
    "shipping_total": 0,
    "shipping_subtotal": 0,
    "shipping_tax_total": 0,
    "original_shipping_tax_total": 0,
    "original_shipping_subtotal": 0,
    "original_shipping_total": 0,
    "credit_line_subtotal": 0,
    "credit_line_tax_total": 0,
    "credit_line_total": 0,
    "metadata": null,
    "sales_channel_id": "sc_01JP5CGAA7THA9DR0JJ8QP3A34",
    "shipping_address_id": "caaddr_01K06YQV61XYRX5JWBE6T095HS",
    "billing_address_id": "caaddr_01K06YQV61CMZ9HSSS4JCEDGWF",
    "customer_id": "cus_01K06YQV5ARK4Y9QXMQ341MZ1B",
    "items": [
        {
            "id": "cali_01K06YQFW99F180CQJWZ1YVG8N",
            "quantity": 1,
            "unit_price": 119,
            "compare_at_unit_price": null,
            "is_tax_inclusive": true,
            "tax_lines": [
                {
                    "id": "calitxl_01K06YR078SK6MBSJ45Z643084",
                    "description": "MwSt",
                    "code": "",
                    "rate": 19,
                    "provider_id": "system"
                }
            ],
            "adjustments": [
                {
                    "id": "caliadj_01K06YR08PWAM10ZACX61MNESR",
                    "code": "FLY119",
                    "promotion_id": "promo_01JZJTKMBZ4PE46QBSJ59MQ8TV",
                    "amount": 100,
                    "is_tax_inclusive": true
                }
            ],
            "product": {
                "id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
                "collection_id": null,
                "type_id": null,
                "categories": [
                    {
                        "id": "pcat_01JP5NSGYQG76M8E68WV13Y4EF"
                    }
                ],
            }
        }
    ],
    "promotions": [
        {
            "id": "promo_01JZJTKMBZ4PE46QBSJ59MQ8TV",
            "code": "FLY119",
            "is_automatic": false,
            "application_method": {
                "value": 119,
                "type": "fixed",
                "currency_code": "eur"
            }
        }
    ]
}
```

scherddel committed:
This fixes the adjustment to be handled as a subtotal value in every calculation and applies the tax inclusive logic on the promotion value itself

scherddel commented:
I also cleaned up some of the variable names and cleaned up a little bit to increase readbility of the code.

scherddel commented:
@fPolic Can you check if it also works with your testcases? I am wondering why this has not popped up earlier which means that the testcases have some mistakes.

scherddel commented:
/snapshot-this

scherddel commented:
Should fix:
- https://github.com/medusajs/medusa/issues/12690
- https://github.com/medusajs/medusa/issues/12226

fPolic commented:
hey @scherddel,
I appreciate your efforts around tax inclusive discounts. Could you describe the scenario in which the `is_tax_inclusive` flag wasn't considered (when any total field is requested the entire adjustment object, including the flag should be passed)?

Regarding PR, the refactor broke a few test cases that should be updated before we can review it. Also, could I get you to add an integration test for the scenario that you described (119 priced item)?

scherddel commented:
hey @fPolic,

let me try to explain where I think it broke because it didnt consider the is_tax_inclusive flag:
packages/core/utils/src/totals/promotion/index.ts
https://github.com/medusajs/medusa/blob/1ba316a26ea6b253a87bbbc9076229550de84272/packages/modules/promotion/src/utils/compute-actions/line-items.ts#L133
```ts
    const amount = calculateAdjustmentAmountFromPromotion(
      item,
      {
        value: promotionValue,
        applied_value: appliedPromoValue,
        max_quantity: maxQuantity,
        type: applicationMethod?.type!,
        allocation,
      },
      lineItemsTotal
    )
```
This returns the value for the actions computed to apply the promotion to the item. If you follow the flow the function that actually returns the promotionvalue and later compares it is
https://github.com/medusajs/medusa/blob/1ba316a26ea6b253a87bbbc9076229550de84272/packages/core/utils/src/totals/promotion/index.ts#L142C3-L146C4
```ts
const promotionValue = getPromotionValue(
    promotion,
    applicableTotal,
    lineItemsTotal
  )
```
This function will always return the promotions TOTAL and not the subtotal (as if does not consider the is_tax_inclusive flag of the promotion)

**This itself would not be the problem alone.**

If you then follow the values that the returned promotionValue is used to compare it will always be compared against a value "appicableTotal" e.g here:
https://github.com/medusajs/medusa/blob/1ba316a26ea6b253a87bbbc9076229550de84272/packages/core/utils/src/totals/promotion/index.ts#L148
or here
https://github.com/medusajs/medusa/blob/1ba316a26ea6b253a87bbbc9076229550de84272/packages/core/utils/src/totals/promotion/index.ts#L105

but applicableTotal is not actually the TOTAL but calculated based on the subtotal of the item:
where applicableTotal =  unitPrice = MathBN.div(lineItem.subtotal, lineItem.quantity) or applicableTotal = lineItemTotal = MathBN.mult(getLineItemUnitPrice(lineItem), quantity) => MathBN.div(lineItem.subtotal, lineItem.quantity)

To me it brakes because you cant use a subtotal based item value and compare it with the total based promotion.

This is why my fix concentrates on adjusting getPromotionValue to consider the is_tax_inclusive flag and return the promotion value in its subtotal so that all subsequent calculations make sense.

I will also work on another test case.

scherddel commented:
@fPolic One extra adjustment that need to be made was the logic around the calculations of the totals for the adjustments. Since the adjustment itself is now reflected as a subtotal the logic here needed to be adjusted as well as we now dont have to consider the tax inclusiveness here:
https://github.com/medusajs/medusa/blob/1ba316a26ea6b253a87bbbc9076229550de84272/packages/core/utils/src/totals/adjustment/index.ts#L39

scherddel committed:
Added some testcases and revoked some changes to improve testing output

scherddel commented:
@fPolic I added some testcases and I had to correct the ones from you for the cart calculation as some values were wrong.  You can have a look at let me know what you think. 

<img width="896" height="316" alt="image" src="https://github.com/user-attachments/assets/bf2774d8-c82d-4a91-ac8b-10e50527163e" />

Until now I have not seen any issue to make adjustments subtotals per default. So my PR should work.

Unfortunetaly the test cases for admin/promotion do not really work on my end so I cant really work on them. Do you have a similar issue when running yarn run test:integration:http?

<img width="1392" height="536" alt="image" src="https://github.com/user-attachments/assets/a28ae97d-d2ee-48b5-b507-374f08b95c7c" />

fPolic commented:
> Unfortunetaly the test cases for admin/promotion do not really work on my end so I cant really work on them. Do you have a similar issue when running yarn run test:integration:http?

From the screenshot, tests are working properly but failing because the received value is different from what is expected.

fPolic commented on packages/core/utils/src/totals/__tests__/totals.ts:
```diff
@@ -595,51 +594,50 @@ describe("Total calculation", function () {
       ],
       subtotal: 100,
-      tax_total: 18.333333333333332,
-      total: 110, // total is 120 - 10 tax inclusive discount
```
todo: this changed the point of this test, if you look at the comment above the test, we have:
- 120$ original total
- 10$ fixed discount 
- tax inclusive context & promotion

So we want the total to be 120$ - 10$ = 110$ here.

scherddel replied:
I dont think your feedback is really constructive. If you want to point out a mistake I made it is worth looking into what you are trying to achieve. 
We can either change the comment or the value so that the total of 110 is met. I adjusted the comment and picked the correct adjustment amount so the total becomes 110.

scherddel commented:
> > Unfortunetaly the test cases for admin/promotion do not really work on my end so I cant really work on them. Do you have a similar issue when running yarn run test:integration:http?
> 
> 
> 
> From the screenshot, tests are working properly but failing because the received value is different from what is expected.

Yes but this also fails with fresh pull without any of my adjustments. I dont think this has to do with my setup but the test itself brakes..

scherddel committed:
Fixed a test case based on feedback

scherddel commented:
I think i fixed the test case execution for me. Somehow my local branch did not fully take all commits into consideration

scherddel committed:
Corrected promotion/admin test cases

scherddel committed:
Corrected cart/store test case

scherddel commented:
Also corrected the last test case now. All seems to work now:
<img width="928" height="350" alt="image" src="https://github.com/user-attachments/assets/1498cb8a-a513-4487-83bc-d63976a89deb" />

scherddel committed:
Improved cart/store test cases for more robust promotion testing considering tax inclusion flags

scherddel commented:
@fPolic I also added a couple of new test cases that also cover the use cases provided in my inital observation

scherddel committed:
Remove unnessary changes as adjustments now automatically are subtotals and therefore the tax inclusive flag does not need to be applied again

scherddel commented:
I think we should discuss whether the is_tax_inclusive field is actually needed in adjustments at this point. Since an adjustment now always represents the subtotal of the promotion, this flag doesn't add any value here.
I'd recommend removing this field from adjustments and keeping it only in the promotion object (having it in both places creates unnecessary redundancy anyway).

What are your thoughts on this approach?

scherddel committed:
Remove adjustments->is_tax_inclusive usage everywhere

scherddel committed:
Migration script to remove is_tax_inclusive in cart line item adjustment

scherddel commented:
I've implemented the complete removal of is_tax_inclusive from adjustments. Please take a look and let me know if this aligns with your thoughts and if you're comfortable with the approach.

scherddel committed:
Forgot to adjust one more testcase

fPolic commented on integration-tests/http/__tests__/cart/store/cart.spec.ts:
```diff
@@ -16,6 +16,7 @@ import {
 import { setupTaxStructure } from "../../../../modules/__tests__/fixtures"
 import { createAuthenticatedCustomer } from "../../../../modules/helpers/create-authenticated-customer"
 import { medusaTshirtProduct } from "../../../__fixtures__/product"
+import { application } from "express"
```
todo: remove

fPolic commented on integration-tests/http/__tests__/cart/store/cart.spec.ts:
```diff
@@ -2810,6 +2811,238 @@ medusaIntegrationTestRunner({
+        })
+
+        it("should add a 105 USD tax inclusive promotion for a 105 USD tax inclusive item and logically result in a 0 total with tax 5%", async () => {
+          const taxExclPromotion = (
```
nit: `taxInclPromotion`

fPolic commented on packages/modules/cart/src/migrations/Migration20250717142300.ts:
```diff
@@ -0,0 +1,15 @@
+export class Migration20250717142300 extends Migration {
+  async up(): Promise<void> {
+    this.addSql(
+      'alter table if exists "cart_line_item_adjustment" drop column if exists "is_tax_inclusive";'
```
todo: if we decide to remove the flag, we would also need to remove it from the order line item adjustment

fPolic commented on packages/core/utils/src/totals/promotion/index.ts:
```diff
@@ -87,22 +121,34 @@ export function calculateAdjustmentAmountFromPromotion(
+  const sumTax = MathBN.sum(
+    ...((lineItem.tax_lines ?? []).map((taxLine) => taxLine.rate) ?? [])
+  )
+  const sumTaxRate = MathBN.div(sumTax, 100)
```
discussion: concern here is if the promotion module should handle tax details like this. It can be considered as a responsibility of another module but also as a part of the calculation context of the promotion so would like to hear what others think here as well.
cc @olivermrbl

fPolic reviewed (commented):
Tests look good now @scherddel!
There is a discussion point I would like to go over with the team.

scherddel committed:
Corrections based on fPolic feedback

olivermrbl commented:
@scherddel, just wanted to give you an update. We are discussing this change internally and expect to get back with a review later this week :)

olivermrbl commented:
Hey @scherddel, I gave all the comments and examples in this PR a read, and tbqh, I have a hard time understanding what this PR is aimed at solving. There has been a lot of back and forth, and I believe we also integrated fixes from other PRs that solved a few of the example scenarios, that you have described.

So could I please get you to describe precisely what the issue is, so I can have that in mind while reviewing the PR?

And even better, if you could also describe exact reproducible steps for the issue.

Thanks!

scherddel commented:
@olivermrbl Thanks for coming back to me. I will try to guide you through the issue again. It could be a longer post. As a reference, I updated my local instance to the latest release 2.8.8. 

1. Add the following product into your cart (copied from items output) and make sure the cart has a shipping address to set tax_lines properly
```
{
    "id": "cali_01K117BJ6KGG8RY5S049G2C512",
    "variant_id": "variant_01JTQHC1M3A9XEKX1QEVX691JK",
    "product_id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
    "requires_shipping": false,
    "quantity": 1,
    "unit_price": 119,
    "is_tax_inclusive": true,
    "tax_lines": [
        {
            "id": "calitxl_01K117BJAFSFD16KMT1J4BFZJE",
            "description": "MwSt",
            "code": "",
            "rate": 19,
            "provider_id": "system"
        }
    ],
    "adjustments": [],
}
```
2. Cart looks as follows (removed unnessary fields, no suprises, subtotal, total and tax are calculated correctly)
```
{
    "id": "cart_01K10X4MHYEXRF1CCCTRHW8R6Y",
    "currency_code": "eur",
    "email": "test@test.com",
    "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
    "created_at": "2025-07-25T13:42:39.422Z",
    "updated_at": "2025-07-25T13:43:18.271Z",
    "completed_at": null,
    "total": 119,
    "subtotal": 100,
    "tax_total": 19,
    "discount_total": 0,
    "discount_subtotal": 0,
    "discount_tax_total": 0,
    "original_total": 119,
    "original_tax_total": 19,
    "item_total": 119,
    "item_subtotal": 100,
    "item_tax_total": 19,
    "original_item_total": 119,
    "original_item_subtotal": 100,
    "original_item_tax_total": 19,
    "shipping_total": 0,
    "shipping_subtotal": 0,
    "shipping_tax_total": 0,
    "original_shipping_tax_total": 0,
    "original_shipping_subtotal": 0,
    "original_shipping_total": 0,
    "credit_line_subtotal": 0,
    "credit_line_tax_total": 0,
    "credit_line_total": 0,
    "metadata": {
    },
    "sales_channel_id": "sc_01JP5CGAA7THA9DR0JJ8QP3A34",
    "shipping_address_id": "caaddr_01K10X5NSHDZNY05TS53B47MN2",
    "billing_address_id": "caaddr_01K10X5NSH02WG66CJHHS8CFZ6",
    "customer_id": "cus_01JQ4944V31GBEPF64GYETNEWR",
    "items": [
        {
            "id": "cali_01K117BJ6KGG8RY5S049G2C512",
            "variant_id": "variant_01JTQHC1M3A9XEKX1QEVX691JK",
            "product_id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
            "quantity": 1,
            "unit_price": 119,
            "compare_at_unit_price": null,
            "is_tax_inclusive": true,
            "tax_lines": [
                {
                    "id": "calitxl_01K117BJAFSFD16KMT1J4BFZJE",
                    "description": "MwSt",
                    "code": "",
                    "rate": 19,
                    "provider_id": "system"
                }
            ],
            "adjustments": [],
            "product": {
                "id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
                "collection_id": null,
                "type_id": null,
                "categories": [
                    {
                        "id": "pcat_01JP5NSGYQG76M8E68WV13Y4EF"
                    }
                ],
                "tags": [
                    {
                        "id": "ptag_01JTE5VXQVH6CAWK3G4HBDH9W3"
                    },
                    {
                        "id": "ptag_01JTE5W3N65N5V6RB8D38H7FEJ"
                    }
                ]
            }
        }
    ],
    "shipping_methods": [
        {
            "amount": 0,
            "is_tax_inclusive": true,
            "shipping_option_id": "so_01JV9W0WQ3NH5VAN8R95SYC7K9",
            "name": "Email",
            "id": "casm_01K10X5QNK18XWYPZAZQ9NJ775",
            "tax_lines": [
                {
                    "id": "casmtxl_01K10X5TNDZ2MY49YFXBZ3SS1M",
                    "description": "MwSt",
                    "code": "",
                    "rate": 19,
                    "provider_id": "system"
                }
            ],
            "adjustments": []
        }
    ],
    "shipping_address": {
        "id": "caaddr_01K10X5NSHDZNY05TS53B47MN2",
        "first_name": "ads",
        "last_name": "ads",
        "company": "",
        "address_1": "asd",
        "address_2": "",
        "city": "asd",
        "postal_code": "324",
        "country_code": "de",
        "province": null,
        "phone": ""
    },
    "billing_address": {
        "id": "caaddr_01K10X5NSH02WG66CJHHS8CFZ6",
        "first_name": "ads",
        "last_name": "ads",
        "company": "",
        "address_1": "asd",
        "address_2": "",
        "city": "asd",
        "postal_code": "324",
        "country_code": "de",
        "province": null,
        "phone": ""
    },
    "credit_lines": [],
    "customer": {
        "id": "cus_01JQ4944V31GBEPF64GYETNEWR",
        "email": "test@test.com",
        "groups": []
    },
    "region": {
        "id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
        "name": "Europe",
        "currency_code": "eur",
        "automatic_taxes": true,
        "countries": [
            {
                "iso_2": "at",
                "iso_3": "aut",
                "num_code": "040",
                "name": "AUSTRIA",
                "display_name": "Austria",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.441Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            },
            {
                "iso_2": "ch",
                "iso_3": "che",
                "num_code": "756",
                "name": "SWITZERLAND",
                "display_name": "Switzerland",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.443Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            },
            {
                "iso_2": "de",
                "iso_3": "deu",
                "num_code": "276",
                "name": "GERMANY",
                "display_name": "Germany",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.441Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            },
            {
                "iso_2": "lu",
                "iso_3": "lux",
                "num_code": "442",
                "name": "LUXEMBOURG",
                "display_name": "Luxembourg",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.442Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            }
        ]
    },
    "promotions": [],
    "payment_collection": {
        "id": "pay_col_01K10X5QWGC5E21PJFNZXACHFM",
        "currency_code": "eur",
        "completed_at": null,
        "status": "not_paid",
        "metadata": null,
        "raw_amount": {
            "value": "119",
            "precision": 20
        },
        "raw_authorized_amount": null,
        "raw_captured_amount": null,
        "raw_refunded_amount": null,
        "created_at": "2025-07-25T13:43:15.601Z",
        "updated_at": "2025-07-25T16:41:15.123Z",
        "deleted_at": null,
        "payment_sessions": [],
        "amount": 119,
        "authorized_amount": null,
        "captured_amount": null,
        "refunded_amount": null
    }
}
```
3. Add a promotion that looks as follows
```
{
  "id": "promo_01K117B6AJ7DFWFF8DXNXG63DS",
  "code": "TAXINCL119",
  "is_automatic": false,
  "is_tax_inclusive": true,
  "type": "standard",
  "status": "active",
  "created_at": "2025-07-25T16:40:59.986Z",
  "updated_at": "2025-07-25T16:40:59.986Z",
  "deleted_at": null,
  "campaign_id": null,
  "application_method": {
    "id": "proappmet_01K117B6AJN43FP5S3Q1NKZMFS",
    "currency_code": "eur",
    "max_quantity": 1,
    "apply_to_quantity": null,
    "buy_rules_min_quantity": null,
    "type": "fixed",
    "target_type": "items",
    "allocation": "each",
    "raw_value": {
      "value": "119",
      "precision": 20
    },
    "created_at": "2025-07-25T16:40:59.986Z",
    "updated_at": "2025-07-25T16:40:59.986Z",
    "deleted_at": null,
    "promotion_id": "promo_01K117B6AJ7DFWFF8DXNXG63DS",
    "target_rules": [],
    "buy_rules": [],
    "value": 119
  },
  "rules": [],
  "campaign": null
}

```
4. Cart will look like this 
discount_subtotal=100 -> CORRECT
discount_tax_total=15,966 -> WRONG
discount_total=100 -> WRONG

adjustment=100 -> could be considered WRONG, I will come to this in my next proof.
```
{
    "id": "cart_01K10X4MHYEXRF1CCCTRHW8R6Y",
    "currency_code": "eur",
    "email": "test@test.com",
    "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
    "created_at": "2025-07-25T13:42:39.422Z",
    "updated_at": "2025-07-25T13:43:18.271Z",
    "completed_at": null,
    "total": 0,
    "subtotal": 100,
    "tax_total": 0,
    "discount_total": 100,
    "discount_subtotal": 100,
    "discount_tax_total": 15.966386554621849,
    "original_total": 119,
    "original_tax_total": 19,
    "item_total": 0,
    "item_subtotal": 100,
    "item_tax_total": 0,
    "original_item_total": 119,
    "original_item_subtotal": 100,
    "original_item_tax_total": 19,
    "shipping_total": 0,
    "shipping_subtotal": 0,
    "shipping_tax_total": 0,
    "original_shipping_tax_total": 0,
    "original_shipping_subtotal": 0,
    "original_shipping_total": 0,
    "credit_line_subtotal": 0,
    "credit_line_tax_total": 0,
    "credit_line_total": 0,
    "metadata": {
    },
    "sales_channel_id": "sc_01JP5CGAA7THA9DR0JJ8QP3A34",
    "shipping_address_id": "caaddr_01K10X5NSHDZNY05TS53B47MN2",
    "billing_address_id": "caaddr_01K10X5NSH02WG66CJHHS8CFZ6",
    "customer_id": "cus_01JQ4944V31GBEPF64GYETNEWR",
    "items": [
        {
            "id": "cali_01K117BJ6KGG8RY5S049G2C512",
            "variant_id": "variant_01JTQHC1M3A9XEKX1QEVX691JK",
            "product_id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
            "requires_shipping": false,
            "metadata": {},
            "created_at": "2025-07-25T16:41:12.147Z",
            "updated_at": "2025-07-25T16:41:12.147Z",
            "quantity": 1,
            "unit_price": 119,
            "compare_at_unit_price": null,
            "is_tax_inclusive": true,
            "tax_lines": [
                {
                    "id": "calitxl_01K117XT5N5HTZ08CQM4JT5RDZ",
                    "description": "MwSt",
                    "code": "",
                    "rate": 19,
                    "provider_id": "system"
                }
            ],
            "adjustments": [
                {
                    "id": "caliadj_01K117XT7B7KEV1DZS09MP5FCE",
                    "code": "TAXINCL119",
                    "promotion_id": "promo_01K117B6AJ7DFWFF8DXNXG63DS",
                    "amount": 100
                }
            ],
            "product": {
                "id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
                "collection_id": null,
                "type_id": null,
                "categories": [
                    {
                        "id": "pcat_01JP5NSGYQG76M8E68WV13Y4EF"
                    }
                ],
                "tags": [
                    {
                        "id": "ptag_01JTE5VXQVH6CAWK3G4HBDH9W3"
                    },
                    {
                        "id": "ptag_01JTE5W3N65N5V6RB8D38H7FEJ"
                    }
                ]
            }
        }
    ],
    "shipping_methods": [
        {
            "amount": 0,
            "is_tax_inclusive": true,
            "shipping_option_id": "so_01JV9W0WQ3NH5VAN8R95SYC7K9",
            "name": "Email",
            "id": "casm_01K10X5QNK18XWYPZAZQ9NJ775",
            "tax_lines": [
                {
                    "id": "casmtxl_01K117XT5PA3B13FMD31JRXNEE",
                    "description": "MwSt",
                    "code": "",
                    "rate": 19,
                    "provider_id": "system"
                }
            ],
            "adjustments": []
        }
    ],
    "shipping_address": {
        "id": "caaddr_01K10X5NSHDZNY05TS53B47MN2",
        "first_name": "ads",
        "last_name": "ads",
        "company": "",
        "address_1": "asd",
        "address_2": "",
        "city": "asd",
        "postal_code": "324",
        "country_code": "de",
        "province": null,
        "phone": ""
    },
    "billing_address": {
        "id": "caaddr_01K10X5NSH02WG66CJHHS8CFZ6",
        "first_name": "ads",
        "last_name": "ads",
        "company": "",
        "address_1": "asd",
        "address_2": "",
        "city": "asd",
        "postal_code": "324",
        "country_code": "de",
        "province": null,
        "phone": ""
    },
    "credit_lines": [],
    "customer": {
        "id": "cus_01JQ4944V31GBEPF64GYETNEWR",
        "email": "test@test.com",
        "groups": []
    },
    "region": {
        "id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
        "name": "Europe",
        "currency_code": "eur",
        "automatic_taxes": true,
        "countries": [
            {
                "iso_2": "at",
                "iso_3": "aut",
                "num_code": "040",
                "name": "AUSTRIA",
                "display_name": "Austria",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.441Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            },
            {
                "iso_2": "ch",
                "iso_3": "che",
                "num_code": "756",
                "name": "SWITZERLAND",
                "display_name": "Switzerland",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.443Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            },
            {
                "iso_2": "de",
                "iso_3": "deu",
                "num_code": "276",
                "name": "GERMANY",
                "display_name": "Germany",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.441Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            },
            {
                "iso_2": "lu",
                "iso_3": "lux",
                "num_code": "442",
                "name": "LUXEMBOURG",
                "display_name": "Luxembourg",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.442Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            }
        ]
    },
    "promotions": [
        {
            "id": "promo_01K117B6AJ7DFWFF8DXNXG63DS",
            "code": "TAXINCL119",
            "is_automatic": false,
            "application_method": {
                "value": 119,
                "type": "fixed",
                "currency_code": "eur"
            }
        }
    ],
    "payment_collection": {
        "id": "pay_col_01K10X5QWGC5E21PJFNZXACHFM",
        "currency_code": "eur",
        "completed_at": null,
        "status": "not_paid",
        "metadata": null,
        "raw_amount": {
            "value": "19",
            "precision": 20
        },
        "raw_authorized_amount": null,
        "raw_captured_amount": null,
        "raw_refunded_amount": null,
        "created_at": "2025-07-25T13:43:15.601Z",
        "updated_at": "2025-07-25T16:51:10.207Z",
        "deleted_at": null,
        "payment_sessions": [],
        "amount": 19,
        "authorized_amount": null,
        "captured_amount": null,
        "refunded_amount": null
    }
}
```

As you can see discount values are completely off and do not reflect the correct values considering the tax_inclusive flag. (In my logic if a promotion is tax_inclusive then the discount_total=119)

At least this one already shows inconsitencies that are corrected with my fix and obviously need to be fixed. 

I will post a subsequent post with another example using the same data that exposes another issue.

scherddel commented:
@olivermrbl continue here

1. Added another product to the cart with the following prices (still tax_inclusive)
```
{
    "id": "cali_01K118H256H1DSAYCKVZSGDRX2",
    "variant_id": "variant_01JTQHN0ZSM5RXX64E17K8BY7W",
    "product_id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
    "requires_shipping": false,
    "metadata": {},
    "created_at": "2025-07-25T17:01:40.902Z",
    "updated_at": "2025-07-25T17:01:40.902Z",
    "quantity": 1,
    "unit_price": 1190,
    "compare_at_unit_price": null,
    "is_tax_inclusive": true,
    "tax_lines": [
        {
            "id": "calitxl_01K118H289MVMAMGVXT1RDAE97",
            "description": "MwSt",
            "code": "",
            "rate": 19,
            "provider_id": "system"
        }
    ],
    "adjustments": [],
    "product": {
        "id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
        "collection_id": null,
        "type_id": null,
        "categories": [
            {
                "id": "pcat_01JP5NSGYQG76M8E68WV13Y4EF"
            }
        ],
        "tags": [
            {
                "id": "ptag_01JTE5VXQVH6CAWK3G4HBDH9W3"
            },
            {
                "id": "ptag_01JTE5W3N65N5V6RB8D38H7FEJ"
            }
        ]
    }
}
```
2. Apply same promotion as before (119 tax incl)
discount_subtotal=119 -> WRONG
discount_tax_total=15,966 -> WRONG
discount_total=119 -> CORRECT
adjustment=119 -> now the adjustment is 119 which seems unlogical compared to the example before as it now reflects the total value of the promotion and therefore the calculation is not consistent (same promotion)

Now where it gets even worse is if you look at the other totals:
total=1048.39 -> WRONG as it has to be 1071
item_total=1048.39 -> WRONG
item_tax_total= 167.39 -> WRONG
tax_total = 167.39 -> WRONG as it should be 171
```
{
    "id": "cart_01K10X4MHYEXRF1CCCTRHW8R6Y",
    "currency_code": "eur",
    "email": "test@test.com",
    "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
    "created_at": "2025-07-25T13:42:39.422Z",
    "updated_at": "2025-07-25T13:43:18.271Z",
    "completed_at": null,
    "total": 1048.39,
    "subtotal": 1000,
    "tax_total": 167.39,
    "discount_total": 119,
    "discount_subtotal": 119,
    "discount_tax_total": 19,
    "original_total": 1190,
    "original_tax_total": 190,
    "item_total": 1048.39,
    "item_subtotal": 1000,
    "item_tax_total": 167.39,
    "original_item_total": 1190,
    "original_item_subtotal": 1000,
    "original_item_tax_total": 190,
    "shipping_total": 0,
    "shipping_subtotal": 0,
    "shipping_tax_total": 0,
    "original_shipping_tax_total": 0,
    "original_shipping_subtotal": 0,
    "original_shipping_total": 0,
    "credit_line_subtotal": 0,
    "credit_line_tax_total": 0,
    "credit_line_total": 0,
    "metadata": {
    },
    "sales_channel_id": "sc_01JP5CGAA7THA9DR0JJ8QP3A34",
    "shipping_address_id": "caaddr_01K10X5NSHDZNY05TS53B47MN2",
    "billing_address_id": "caaddr_01K10X5NSH02WG66CJHHS8CFZ6",
    "customer_id": "cus_01JQ4944V31GBEPF64GYETNEWR",
    "items": [
        {
            "id": "cali_01K118H256H1DSAYCKVZSGDRX2",
            "variant_id": "variant_01JTQHN0ZSM5RXX64E17K8BY7W",
            "product_id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
            "quantity": 1,
            "unit_price": 1190,
            "compare_at_unit_price": null,
            "is_tax_inclusive": true,
            "tax_lines": [
                {
                    "id": "calitxl_01K118KMYHSZD759AXGVJV294V",
                    "description": "MwSt",
                    "code": "",
                    "rate": 19,
                    "provider_id": "system"
                }
            ],
            "adjustments": [
                {
                    "id": "caliadj_01K118KMZWGNJ2SEVVMZESYXK7",
                    "code": "TAXINCL119",
                    "promotion_id": "promo_01K117B6AJ7DFWFF8DXNXG63DS",
                    "amount": 119
                }
            ],
            "product": {
                "id": "prod_01JP5NRZXYXQ88G08D6YNB7BE2",
                "collection_id": null,
                "type_id": null,
                "categories": [
                    {
                        "id": "pcat_01JP5NSGYQG76M8E68WV13Y4EF"
                    }
                ],
                "tags": [
                    {
                        "id": "ptag_01JTE5VXQVH6CAWK3G4HBDH9W3"
                    },
                    {
                        "id": "ptag_01JTE5W3N65N5V6RB8D38H7FEJ"
                    }
                ]
            }
        }
    ],
    "shipping_methods": [
        {
            "amount": 0,
            "is_tax_inclusive": true,
            "shipping_option_id": "so_01JV9W0WQ3NH5VAN8R95SYC7K9",
            "name": "Email",
            "id": "casm_01K10X5QNK18XWYPZAZQ9NJ775",
            "tax_lines": [
                {
                    "id": "casmtxl_01K118KMYJWRFMYTJD90MHCDKE",
                    "description": "MwSt",
                    "code": "",
                    "rate": 19,
                    "provider_id": "system"
                }
            ],
            "adjustments": []
        }
    ],
    "shipping_address": {
        "id": "caaddr_01K10X5NSHDZNY05TS53B47MN2",
        "first_name": "ads",
        "last_name": "ads",
        "company": "",
        "address_1": "asd",
        "address_2": "",
        "city": "asd",
        "postal_code": "324",
        "country_code": "de",
        "province": null,
        "phone": ""
    },
    "billing_address": {
        "id": "caaddr_01K10X5NSH02WG66CJHHS8CFZ6",
        "first_name": "ads",
        "last_name": "ads",
        "company": "",
        "address_1": "asd",
        "address_2": "",
        "city": "asd",
        "postal_code": "324",
        "country_code": "de",
        "province": null,
        "phone": ""
    },
    "credit_lines": [],
    "customer": {
        "id": "cus_01JQ4944V31GBEPF64GYETNEWR",
        "email": "test@test.com",
        "groups": []
    },
    "region": {
        "id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
        "name": "Europe",
        "currency_code": "eur",
        "automatic_taxes": true,
        "countries": [
            {
                "iso_2": "at",
                "iso_3": "aut",
                "num_code": "040",
                "name": "AUSTRIA",
                "display_name": "Austria",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.441Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            },
            {
                "iso_2": "ch",
                "iso_3": "che",
                "num_code": "756",
                "name": "SWITZERLAND",
                "display_name": "Switzerland",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.443Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            },
            {
                "iso_2": "de",
                "iso_3": "deu",
                "num_code": "276",
                "name": "GERMANY",
                "display_name": "Germany",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.441Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            },
            {
                "iso_2": "lu",
                "iso_3": "lux",
                "num_code": "442",
                "name": "LUXEMBOURG",
                "display_name": "Luxembourg",
                "region_id": "reg_01JP5CGCC1HN1Y2R2CHNEGZEDZ",
                "metadata": null,
                "created_at": "2025-03-12T14:35:25.442Z",
                "updated_at": "2025-03-24T19:07:44.743Z",
                "deleted_at": null
            }
        ]
    },
    "promotions": [
        {
            "id": "promo_01K117B6AJ7DFWFF8DXNXG63DS",
            "code": "TAXINCL119",
            "is_automatic": false,
            "application_method": {
                "value": 119,
                "type": "fixed",
                "currency_code": "eur"
            }
        }
    ],
    "payment_collection": {
        "id": "pay_col_01K10X5QWGC5E21PJFNZXACHFM",
        "currency_code": "eur",
        "completed_at": null,
        "status": "not_paid",
        "metadata": null,
        "raw_amount": {
            "value": "1071",
            "precision": 20
        },
        "raw_authorized_amount": null,
        "raw_captured_amount": null,
        "raw_refunded_amount": null,
        "created_at": "2025-07-25T13:43:15.601Z",
        "updated_at": "2025-07-25T17:03:05.740Z",
        "deleted_at": null,
        "payment_sessions": [],
        "amount": 1071,
        "authorized_amount": null,
        "captured_amount": null,
        "refunded_amount": null
    }
}
```

As you can see there are obviously critical failures in the calculation that are all fixed by my PR. Let me know if you need anything else.

scherddel commented:
Hi @olivermrbl,
I hope you're doing well! I wanted to respectfully follow up on this PR as it addresses some critical issues for our project. I've invested significant time debugging and testing to ensure this resolves all the outstanding concerns that were previously raised.
I understand there have been questions about the root cause in earlier discussions, which understandably led to some delays. I believe this implementation now properly addresses those issues, but I'm absolutely open to discussing any remaining concerns you might have.
Given the impact this has on our project timeline, I'd greatly appreciate your feedback at your earliest convenience. If there are any additional questions or modifications needed, please let me know and I'll address them promptly.

Thank you for your time and consideration – looking forward to getting this into the next release if possible!
Best regards,

olivermrbl commented:
> Hi @olivermrbl, I hope you're doing well! I wanted to respectfully follow up on this PR as it addresses some critical issues for our project. I've invested significant time debugging and testing to ensure this resolves all the outstanding concerns that were previously raised. I understand there have been questions about the root cause in earlier discussions, which understandably led to some delays. I believe this implementation now properly addresses those issues, but I'm absolutely open to discussing any remaining concerns you might have. Given the impact this has on our project timeline, I'd greatly appreciate your feedback at your earliest convenience. If there are any additional questions or modifications needed, please let me know and I'll address them promptly.
> 
> Thank you for your time and consideration – looking forward to getting this into the next release if possible! Best regards,

@scherddel, thanks for sharing the example scenarios. I do indeed see the issue now and agree that we need to solve this. I didn't get a chance to review the PR today, but I'll make sure to prioritize it tomorrow. Again, really appreciate the time you've put into this.

olivermrbl commented:
/snapshot-this

olivermrbl commented on packages/modules/cart/src/migrations/Migration20250717142300.ts:
```diff
@@ -0,0 +1,15 @@
+import { Migration } from "@mikro-orm/migrations"
+
+export class Migration20250717142300 extends Migration {
```
thought: These migrations are a breaking change; I'd really love to avoid this.

For one, we could have queries that rely on the existence of this field and making this change would break their app.

Secondly, this change to the "calculation process" will also mean that all tax-inclusive orders and carts up until this point will be inconsistent with the new business logic and eventually compute incorrect totals. 

I'd like to suggest we try to solve the "Known issue" assuming the flag will continue to live on. Do you think you'd be up for exploring that?

scherddel replied:
Thank you for the feedback! I'm completely fine with keeping the flag on the adjustment – removing it was primarily motivated by consistency and avoiding redundancy, but it won't impact the core fix.

However, I do want to highlight a potential UX concern: having an adjustment with is_tax_inclusive=true while the value represents only the subtotal could be confusing for developers. To make this approach work, we'd need to align the adjustment logic with how promotions currently handle this (where the promotion amount always reflects the "raw" user input regardless of the flag). This would require a more comprehensive refactoring of the entire adjustment and promotion value application logic – something I was hoping to avoid to keep this PR less radical.

Regarding the calculation process impact: I believe we likely already have incorrect totals in tax-inclusive contexts due to the existing issues. Since most of the changes focus on cart total calculations, this fix would primarily affect existing carts and adjustment values, not completed orders. That said, previously completed orders with promotions are also likely affected since they would have inherited the incorrect totals from their cart state.
I understand the concern about breaking changes, but I'd argue that maintaining the current incorrect implementation to "protect" existing totals might not be the right approach, given that those totals are fundamentally flawed.

Would you be open to discussing which approach you'd prefer? I'm happy to either:
- Keep the current implementation with the flag retained, or
- Explore the more comprehensive refactoring if you think the consistency benefits outweigh the additional complexity

Let me know your thoughts!

scherddel committed:
Refactored PR to consider feedback from oliver

scherddel commented:
Hi @olivermrbl,

I've completed the refactoring based on your feedback. I've not only retained the is_tax_inclusive flag in adjustments but also aligned the adjustment amount logic with how promotion amounts are handled for consistency.
This required some additional restructuring to ensure the logic flows properly across totals. I believe this approach addresses your concerns while maintaining the core fix.

I'd appreciate your review when you have a chance – let me know if this implementation meets your expectations or if there are any other aspects you'd like me to adjust.

Thanks for your patience as we work through this!

olivermrbl commented on packages/core/utils/src/totals/promotion/index.ts:
```diff
@@ -5,23 +5,23 @@ import {
 
-function getPromotionValueForPercentage(promotion, lineItemTotal) {
-  return MathBN.mult(MathBN.div(promotion.value, 100), lineItemTotal)
+function getPromotionValueForPercentage(sumTaxRate, promotion, lineItemAmount) {
```
todo: I believe we can remove the `sumTaxRate`; it's no longer used
```suggestion
function getPromotionValueForPercentage(promotion, lineItemAmount) {
```

olivermrbl commented on packages/core/utils/src/totals/promotion/index.ts:
```diff
@@ -30,16 +30,20 @@ function getPromotionValueForFixed(promotion, itemTotal, allItemsTotal) {
 
-export function getPromotionValue(promotion, lineItemTotal, lineItemsTotal) {
+export function getPromotionValue(
+  sumTaxRate,
```
todo: Remove unused param
```suggestion
```

olivermrbl commented on packages/core/utils/src/totals/promotion/index.ts:
```diff
@@ -30,16 +30,20 @@ function getPromotionValueForFixed(promotion, itemTotal, allItemsTotal) {
+) {
   if (promotion.type === ApplicationMethodType.PERCENTAGE) {
-    return getPromotionValueForPercentage(promotion, lineItemTotal)
+    return getPromotionValueForPercentage(sumTaxRate, promotion, lineItemAmount)
```
```suggestion
    return getPromotionValueForPercentage(promotion, lineItemAmount)
```

olivermrbl commented on packages/core/utils/src/totals/promotion/index.ts:
```diff
@@ -87,22 +95,34 @@ export function calculateAdjustmentAmountFromPromotion(
+  const sumTax = MathBN.sum(
+    ...((lineItem.tax_lines ?? []).map((taxLine) => taxLine.rate) ?? [])
+  )
+  const sumTaxRate = MathBN.div(sumTax, 100)
```
question: Are these still used? I believe if we remove the `sumTaxRate` from `getPromotionValue` these should be redundant computations

olivermrbl reviewed (commented):
Overall, this looks really nice @scherddel. Appreciate this. 

I left a few minor comments, primarily related to clean-up.

Let me cut a snapshot version for testing, when you've addressed those.

scherddel commented:
@olivermrbl Thanks for catching that! Those changes got reintroduced during the refactoring – you know how Git can be sometimes 😅 I've removed them and amended the fix to the latest commit.

Edit: I've also rebased back to v2.8.8 because it somehow failed some unrelated test cases

olivermrbl commented:
No worries @scherddel.

I created a snapshot here: `2.8.9-snapshot-20250730084423`

It would be great if you could also give this a spin. I'll do the same.

Separately, is there a way I can get a hold of you? Wanted to chat about a few things.

scherddel commented:
@olivermrbl Sure, no worries! Sent you an invite on Discord. Let me know if you prefer another way.

olivermrbl committed:
Merge branch 'develop' into fix/correction-adjustment-calculation

scherddel committed:
Merge branch 'medusajs:develop' into fix/correction-adjustment-calculation

scherddel committed:
Merge branch 'develop' into fix/correction-adjustment-calculation

scherddel committed:
Added more testcases for promotion in cart

olivermrbl reviewed (approved):
Great stuff. Really appreciate the contribution here @scherddel <3

olivermrbl commented on packages/core/utils/src/totals/adjustment/index.ts:
```diff
@@ -24,35 +22,22 @@ export function calculateAdjustmentTotal({
 
-      const adjustmentTaxTotal = MathBN.mult(adjustmentSubtotal, taxRate)
-      const adjustmentTotal = MathBN.add(adjustmentSubtotal, adjustmentTaxTotal)
+    adjustmentsSubtotal = MathBN.add(adjustmentsSubtotal, adjustmentSubtotal)
+    adjustmentsTaxTotal = MathBN.add(adjustmentsTaxTotal, adjustmentTaxTotal)
+    adjustmentsTotal = MathBN.add(adjustmentsTotal, adjustmentTotal)
```
nit: are these not redundant assignments?

scherddel replied:
no, they sum up all items adjustments (sub)totals to adjustmentsSubtotal, adjustmentsTotal and adjustmentsTaxTotal (look closely on the wording adjustmentSubtotal -> adjustment!s!Subtotal) :-)

olivermrbl replied:
Ah lol, missed that sorry

scherddel committed:
Merge branch 'medusajs:develop' into fix/correction-adjustment-calculation
